tests/__init__.py

```diff
@@ -26,37 +26,64 @@ def inloop(fn):
             # test_a is executed in the event loop
             pass
 
+        @inloop(setup=spam, teardown=eggs)
+        async def test_b(self):
+            # test_b is executed in the event loop after spam was successfully
+            # executed. When test_b returns, eggs will be executed
+            pass
+
     :param fn: a function to run in the event loop
     :type fn: a function object
 
+    :param setup: a callable to be executed *before* the function
+    :type setup: a callable object
+
+    :param teardown: a callable object to be executed *after* the function
+    :type teardown: a callable object
+
     :return: a decorated function
     :rtype: a function object
+
     """
-    @functools.wraps(fn)
-    def _wrapper(*args, **kwargs):
-        def_loop = asyncio.get_event_loop()
-        tmp_loop = asyncio.new_event_loop()
 
-        # set newly created event loop as default one, so we don't need
-        # to pass it explicitly down the call stack
-        asyncio.set_event_loop(tmp_loop)
+    def decorate(fn):
+        @functools.wraps(fn)
+        def _wrapper(self, *args, **kwargs):
+            def_loop = asyncio.get_event_loop()
+            tmp_loop = asyncio.new_event_loop()
 
-        try:
-            rv = tmp_loop.run_until_complete(fn(*args, **kwargs))
+            # set newly created event loop as default one, so we don't need
+            # to pass it explicitly down the call stack
+            asyncio.set_event_loop(tmp_loop)
 
-            # running a given coroutine may produce another ones to free
-            # resources (e.g. close sockets) when it's done, so we need
-            # to run the event loop one more time to execute them
-            tmp_loop.stop()
-            tmp_loop.run_forever()
-            tmp_loop.close()
+            if setup:
+                setup(self)
```

Comment:
As is, yes, due to the fact we have to pass the ``self`` argument if ``setup`` is a method of a class.

Would you prefer to have an inspect call here to analyze if a single ``self`` argument is expected? (I believe, in Python 3.x there is *no* notion of an unbound method - it will appear as a general function to the outside world).


```python

>>> class Spam:
...     def eggs(self):
...         pass
... 
>>> Spam.eggs
<function Spam.eggs at 0x10c73b378>

>>> def eggs(self):
...     pass
... 
>>> inspect.ismethod(Spam.eggs)
False
>>> inspect.ismethod(eggs)
False
>>> inspect.isfunction(eggs)
True
>>> inspect.isfunction(Spam.eggs)
True

```